File: typescript/tests/relationships_board.test.ts
// relationships_board.test.ts
// Verifies the CircleAI.Relationships port: contacts sorted by name,
// upcoming-this-month by day, last-contact, not-contacted-since.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryRelationshipsBoard,
  RelationshipsDomainContext,
  personContact,
  importantDate,
  contactEvent,
} from "../src/relationships/index";

describe("InMemoryRelationshipsBoard", () => {
  it("lists contacts sorted by name (ordinal)", () => {
    const b = new InMemoryRelationshipsBoard();
    b.addContact(personContact("c1", "Zara", "friend", null));
    b.addContact(personContact("c2", "Ann", "sister", "note"));
    assert.equal(b.getContact("c2")?.relationship, "sister");
    assert.deepEqual(
      b.contacts.map((c) => c.name),
      ["Ann", "Zara"],
    );
  });

  it("lists important dates in the current UTC month ordered by day", () => {
    const b = new InMemoryRelationshipsBoard();
    const now = new Date();
    const y = now.getUTCFullYear();
    const m = now.getUTCMonth();
    const mkUtc = (day: number): Date => new Date(Date.UTC(y, m, day));
    b.addImportantDate(importantDate("d1", "c1", "Birthday", mkUtc(20)));
    b.addImportantDate(importantDate("d2", "c2", "Anniversary", mkUtc(5)));
    // A date in a different month must be excluded.
    b.addImportantDate(importantDate("d3", "c3", "Other", new Date(Date.UTC(y, (m + 1) % 12, 15))));
    assert.deepEqual(
      b.upcomingThisMonth().map((d) => d.dateId),
      ["d2", "d1"],
    );
  });

  it("tracks last contact and flags contacts not reached since a cutoff", () => {
    const b = new InMemoryRelationshipsBoard();
    b.addContact(personContact("c1", "Ann", "friend", null));
    b.addContact(personContact("c2", "Ben", "friend", null)); // never contacted
    b.recordTouchpoint(contactEvent("c1", "call", new Date("2026-01-01T00:00:00Z"), null));
    b.recordTouchpoint(contactEvent("c1", "text", new Date("2026-03-01T00:00:00Z"), "hi"));
    assert.equal(b.lastContact("c1")?.toISOString(), "2026-03-01T00:00:00.000Z");
    assert.equal(b.lastContact("c2"), undefined);
    // Cutoff after c1's last touch → both flagged.
    assert.deepEqual(
      b.notContactedSince(new Date("2026-04-01T00:00:00Z")).map((c) => c.contactId).sort(),
      ["c1", "c2"],
    );
    // Cutoff before c1's last touch → only c2 flagged.
    assert.deepEqual(
      b.notContactedSince(new Date("2026-02-01T00:00:00Z")).map((c) => c.contactId),
      ["c2"],
    );
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(RelationshipsDomainContext.systemPromptSnippet.includes("[DOMAIN: Relationships]"));
    assert.deepEqual(RelationshipsDomainContext.complianceFlags, ["POPIA", "Not_Therapy"]);
    assert.deepEqual(RelationshipsDomainContext.suggestedTools, ["journal", "mood_tracker", "calendar"]);
  });
});

File: go/tests/relationships_board_test.go
// relationships_board_test.go
//
// Verifies the CircleAI.Relationships port (relationships_board.go): contact
// add/get + sorted listing, important dates this month (ordered by day),
// touchpoint tracking with last-contact, and not-contacted-since filtering.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestRelationships_ContactsAndDates(t *testing.T) {
	b := circleai.NewInMemoryRelationshipsBoard()
	b.AddContact(circleai.PersonContact{ContactId: "c1", Name: "Zara", Relationship: "friend"})
	b.AddContact(circleai.PersonContact{ContactId: "c2", Name: "Alex", Relationship: "family"})
	if got, ok := b.GetContact("c1"); !ok || got.Name != "Zara" {
		t.Fatalf("get contact = %+v ok=%v", got, ok)
	}
	all := b.Contacts()
	if len(all) != 2 || all[0].Name != "Alex" || all[1].Name != "Zara" {
		t.Fatalf("contacts sorted by Name failed: %+v", all)
	}

	// Two dates in the current month, added out of day order.
	now := time.Now().UTC()
	d20 := time.Date(now.Year(), now.Month(), 20, 0, 0, 0, 0, time.UTC)
	d05 := time.Date(now.Year(), now.Month(), 5, 0, 0, 0, 0, time.UTC)
	// A date in a different month, which must be excluded.
	other := time.Date(now.Year(), now.Month(), 5, 0, 0, 0, 0, time.UTC).AddDate(0, 6, 0)
	b.AddImportantDate(circleai.ImportantDate{DateId: "x1", ContactId: "c1", Kind: "birthday", Date: d20})
	b.AddImportantDate(circleai.ImportantDate{DateId: "x2", ContactId: "c2", Kind: "anniversary", Date: d05})
	b.AddImportantDate(circleai.ImportantDate{DateId: "x3", ContactId: "c1", Kind: "birthday", Date: other})

	up := b.UpcomingThisMonth()
	if len(up) != 2 || up[0].DateId != "x2" || up[1].DateId != "x1" {
		t.Fatalf("upcoming-this-month ordered by day failed: %+v", up)
	}
}

func TestRelationships_TouchpointsAndStale(t *testing.T) {
	b := circleai.NewInMemoryRelationshipsBoard()
	b.AddContact(circleai.PersonContact{ContactId: "c1", Name: "Zara"})
	b.AddContact(circleai.PersonContact{ContactId: "c2", Name: "Alex"})
	now := time.Now().UTC()
	b.RecordTouchpoint(circleai.ContactEvent{ContactId: "c1", Kind: "call", AtUtc: now.Add(-2 * time.Hour)})
	b.RecordTouchpoint(circleai.ContactEvent{ContactId: "c1", Kind: "text", AtUtc: now.Add(-30 * time.Minute)})

	last, ok := b.LastContact("c1")
	if !ok || !last.Equal(now.Add(-30*time.Minute)) {
		t.Fatalf("last contact = %v ok=%v", last, ok)
	}
	if _, ok := b.LastContact("c2"); ok {
		t.Fatalf("last contact for never-touched must be false")
	}

	// Cutoff one hour ago: c1's last touch (30m ago) is recent, c2 never -> stale.
	stale := b.NotContactedSince(now.Add(-time.Hour))
	if len(stale) != 1 || stale[0].ContactId != "c2" {
		t.Fatalf("not-contacted-since failed: %+v", stale)
	}
}
